File: src/main/java/org/panteleyev/money/app/ContactDialog.java
/*
 Copyright (C) 2017, 2018, 2019, 2020, 2021, 2022 Petr Panteleyev

 This program is free software: you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
 Foundation, either version 3 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with this
 program. If not, see <https://www.gnu.org/licenses/>.
 */
package org.panteleyev.money.app;

import javafx.application.Platform;
import javafx.geometry.VPos;
import javafx.scene.control.ButtonType;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Control;
import javafx.scene.control.TextArea;
import javafx.scene.control.TextField;
import org.controlsfx.validation.ValidationResult;
import org.controlsfx.validation.ValidationSupport;
import org.panteleyev.fx.BaseDialog;
import org.panteleyev.fx.Controller;
import org.panteleyev.money.app.icons.IconManager;
import org.panteleyev.money.model.Contact;
import org.panteleyev.money.model.ContactType;
import org.panteleyev.money.model.Icon;

import java.net.URL;
import java.util.List;
import java.util.UUID;

import static org.panteleyev.fx.FxUtils.COLON;
import static org.panteleyev.fx.FxUtils.fxString;
import static org.panteleyev.fx.LabelFactory.label;
import static org.panteleyev.fx.combobox.ComboBoxBuilder.comboBox;
import static org.panteleyev.fx.grid.GridBuilder.gridCell;
import static org.panteleyev.fx.grid.GridBuilder.gridPane;
import static org.panteleyev.fx.grid.GridRowBuilder.gridRow;
import static org.panteleyev.money.app.GlobalContext.cache;
import static org.panteleyev.money.app.MainWindowController.UI;
import static org.panteleyev.money.app.Styles.GRID_PANE;
import static org.panteleyev.money.app.icons.IconManager.EMPTY_ICON;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_CITY;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_COMMENT;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_CONTACT;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_COUNTRY;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_MOBILE;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_NAME;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_PHONE;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_STREET;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_TYPE;
import static org.panteleyev.money.bundles.Internationalization.I18N_WORD_ZIP;

final class ContactDialog extends BaseDialog<Contact> {
    private final ValidationSupport validation = new ValidationSupport();

    private final ComboBox<ContactType> typeBox = comboBox(ContactType.values(),
            b -> b.withStringConverter(Bundles::translate));
    private final TextField nameField = new TextField();
    private final TextField phoneField = new TextField();
    private final TextField mobileField = new TextField();
    private final TextField emailField = new TextField();
    private final TextField webField = new TextField();
    private final TextArea commentEdit = new TextArea();
    private final TextField streetField = new TextField();
    private final TextField cityField = new TextField();
    private final TextField countryField = new TextField();
    private final TextField zipField = new TextField();
    private final ComboBox<Icon> iconComboBox = new ComboBox<>();

    ContactDialog(Controller owner, URL css, Contact contact) {
        super(owner, css);

        setTitle(fxString(UI, I18N_WORD_CONTACT));

        nameField.setPrefColumnCount(20);
        IconManager.setupComboBox(iconComboBox);

        getDialogPane().setContent(
                gridPane(
                        List.of(
                                gridRow(label(fxString(UI, I18N_WORD_TYPE, COLON)), typeBox, iconComboBox),
                                gridRow(label(fxString(UI, I18N_WORD_NAME, COLON)), gridCell(nameField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_PHONE, COLON)), gridCell(phoneField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_MOBILE, COLON)), gridCell(mobileField, 2, 1)),
                                gridRow(label("E-Mail:"), gridCell(emailField, 2, 1)),
                                gridRow(label("URL:"), gridCell(webField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_STREET, COLON)), gridCell(streetField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_CITY, COLON)), gridCell(cityField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_COUNTRY, COLON)), gridCell(countryField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_ZIP, COLON)), gridCell(zipField, 2, 1)),
                                gridRow(label(fxString(UI, I18N_WORD_COMMENT, COLON)), gridCell(commentEdit, 2, 1))
                                        .withValignment(VPos.TOP)
                        ), b -> b.withStyle(GRID_PANE)
                )
        );

        if (contact != null) {
            typeBox.getSelectionModel().select(contact.type());

            nameField.setText(contact.name());
            phoneField.setText(contact.phone());
            mobileField.setText(contact.mobile());
            emailField.setText(contact.email());
            webField.setText(contact.web());
            commentEdit.setText(contact.comment());
            streetField.setText(contact.street());
            cityField.setText(contact.city());
            countryField.setText(contact.country());
            zipField.setText(contact.zip());
            iconComboBox.getSelectionModel().select(cache().getIcon(contact.iconUuid()).orElse(EMPTY_ICON));
        } else {
            typeBox.getSelectionModel().select(0);
            iconComboBox.getSelectionModel().select(EMPTY_ICON);
        }

        setResultConverter((ButtonType b) -> {
            if (b != ButtonType.OK) {
                return null;
            }

            long now = System.currentTimeMillis();

            var builder = new Contact.Builder(contact)
                    .name(nameField.getText())
                    .type(typeBox.getSelectionModel().getSelectedItem())
                    .phone(phoneField.getText())
                    .mobile(mobileField.getText())
                    .email(emailField.getText())
                    .web(webField.getText())
                    .comment(commentEdit.getText())
                    .street(streetField.getText())
                    .city(cityField.getText())
                    .country(countryField.getText())
                    .zip(zipField.getText())
                    .iconUuid(iconComboBox.getSelectionModel().getSelectedItem().uuid())
                    .modified(now);

            if (contact == null) {
                builder.uuid(UUID.randomUUID())
                        .created(now);
            }

            return builder.build();
        });

        createDefaultButtons(UI, validation.invalidProperty());
        Platform.runLater(this::createValidationSupport);
    }

    private void createValidationSupport() {
        validation.registerValidator(nameField, (Control control, String value) ->
                ValidationResult.fromErrorIf(control, null, value.isEmpty()));

        validation.initInitialDecoration();
    }

    ComboBox<ContactType> getTypeBox() {
        return typeBox;
    }

    TextField getNameField() {
        return nameField;
    }

    TextField getPhoneField() {
        return phoneField;
    }

    TextField getMobileField() {
        return mobileField;
    }

    TextField getEmailField() {
        return emailField;
    }

    TextField getWebField() {
        return webField;
    }

    TextArea getCommentEdit() {
        return commentEdit;
    }

    TextField getStreetField() {
        return streetField;
    }

    TextField getCityField() {
        return cityField;
    }

    TextField getCountryField() {
        return countryField;
    }

    TextField getZipField() {
        return zipField;
    }
}

File: src/main/kotlin/org/panteleyev/money/ContactDialog.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.money

import javafx.application.Platform
import javafx.collections.FXCollections
import javafx.geometry.VPos
import javafx.scene.control.ButtonType
import javafx.scene.control.ChoiceBox
import javafx.scene.control.Control
import javafx.scene.control.Label
import javafx.scene.control.TextArea
import javafx.scene.control.TextField
import javafx.scene.layout.GridPane
import javafx.scene.layout.RowConstraints
import org.controlsfx.validation.ValidationResult
import org.panteleyev.money.persistence.Contact
import org.panteleyev.money.persistence.ContactType
import org.panteleyev.money.persistence.ReadOnlyStringConverter
import org.panteleyev.utilities.fx.BaseDialog
import java.util.ResourceBundle
import java.util.UUID

class ContactDialog(val contact: Contact?) : BaseDialog<Contact>(MainWindowController.CSS_PATH) {
    private val rb = ResourceBundle.getBundle(MainWindowController.UI_BUNDLE_PATH)

    private val typeChoiceBox = ChoiceBox<ContactType>()
    private val nameField = TextField()
    private val phoneField = TextField()
    private val mobileField = TextField()
    private val emailField = TextField()
    private val webField = TextField()
    private val commentEdit = TextArea()
    private val streetField = TextField()
    private val cityField = TextField()
    private val countryField = TextField()
    private val zipField = TextField()

    init {
        title = rb.getString("contact.Dialog.Title")

        dialogPane.content = GridPane().apply {
            styleClass.add(Styles.GRID_PANE)

            var index = 0
            addRow(index++, Label(rb.getString("label.Type")), typeChoiceBox)
            addRow(index++, Label(rb.getString("label.Name")), nameField)
            addRow(index++, Label(rb.getString("label.Phone")), phoneField)
            addRow(index++, Label(rb.getString("label.Mobile")), mobileField)
            addRow(index++, Label(rb.getString("label.Email")), emailField)
            addRow(index++, Label("URL:"), webField)
            addRow(index++, Label(rb.getString("label.Street")), streetField)
            addRow(index++, Label(rb.getString("label.City")), cityField)
            addRow(index++, Label(rb.getString("label.Country")), countryField)
            addRow(index++, Label(rb.getString("label.ZIP")), zipField)
            addRow(index, Label(rb.getString("label.Comment")), commentEdit)

            val topAlignmentConstraints = RowConstraints()
            topAlignmentConstraints.valignment = VPos.TOP
            for (i in 0..index - 1) {
                rowConstraints.add(RowConstraints())
            }
            rowConstraints.add(topAlignmentConstraints)
        }

        nameField.prefColumnCount = 20

        typeChoiceBox.items = FXCollections.observableArrayList(*ContactType.values())
        typeChoiceBox.converter = object : ReadOnlyStringConverter<ContactType>() {
            override fun toString(type: ContactType): String {
                return type.typeName
            }
        }

        if (contact != null) {
            typeChoiceBox.selectionModel.select(contact.type)

            nameField.text = contact.name
            phoneField.text = contact.phone
            mobileField.text = contact.mobile
            emailField.text = contact.email
            webField.text = contact.web
            commentEdit.text = contact.comment
            streetField.text = contact.street
            cityField.text = contact.city
            countryField.text = contact.country
            zipField.text = contact.zip
        } else {
            typeChoiceBox.selectionModel.select(0)
        }

        setResultConverter { b: ButtonType ->
            if (b == ButtonType.OK) {
                return@setResultConverter Contact(contact?.id ?: 0,
                        typeId = typeChoiceBox.selectionModel.selectedItem.id,
                        name = nameField.text,
                        phone = phoneField.text,
                        mobile = mobileField.text,
                        email = emailField.text,
                        web = webField.text,
                        comment = commentEdit.text,
                        street = streetField.text,
                        city = cityField.text,
                        country = countryField.text,
                        zip = zipField.text,
                        guid = contact?.guid ?: UUID.randomUUID().toString(),
                        modified = System.currentTimeMillis()
                )
            } else {
                return@setResultConverter null
            }
        }

        createDefaultButtons(rb)
        Platform.runLater { createValidationSupport() }
    }

    private fun createValidationSupport() {
        validation.registerValidator(nameField) {
            control: Control, value: String ->
            ValidationResult.fromErrorIf(control, null, value.isEmpty())
        }
        validation.initInitialDecoration()
    }

}